server/lib/two-factor-authentication.js

```diff
@@ -22,6 +22,20 @@ export function verifyTwoFactorAuthenticatorCode(encryptedTwoFactorAuthToken, tw
   return verified;
 }
 
+/** Verifies a submitted recovery code against the hashed ones saved for them in the DB.
+ * If the recovery code is valid, it returns the remaining recovery codes to update the user with.
+ * If the recovery code is not valid, it returns null so we can throw an error.
+ */
+export function verifyTwoFactorAuthenticationRecoveryCode(recoveryCode, hashedRecoveryCodes) {
+  const verified = hashedRecoveryCodes.includes(crypto.hash(recoveryCode));
+  if (verified) {
+    return hashedRecoveryCodes.filter(code => crypto.hash(recoveryCode) !== code);
```

Comment:
It's not the job of this function to filter the codes, this should be moved elsewhere.

It should return `true`/`false`. Or alternatively, do nothing/throw an exception.

Reply:
I put it here because I was trying to keep stuff relating to `crypto` in one file (this one), and returned the recovery codes to the other file so that I didn't have to import the `User` model multiple places. There was actually some thought behind the way I did this and I did in fact consider the way you are suggesting now (but went with this way for the mentioned reasons). Happy to change it but it's not from lack of consideration about the issue.

Reply:
Good to understand the reasons behind this decision but I'm standing by my comment, `verifyTwoFactorAuthenticationRecoveryCode` should only verify.

Maybe you can find another design, for example exposing a function that only compare 1 single recovery code against user input, and directly use this function in the filter.